### PR TITLE
TestCluster: wait for cluster stabilization before starting tests

### PR DESCRIPTION
**Pull request by ReubenBond**
This PR adds a check in `TestCluster` to verify that each silo has observed the expected number of active silos

### CODE REVIEW

**Comment by ReubenBond** [comment]
@dotnet-bot test this please

File: src/Orleans.TestingHost/TestCluster.cs [hunk 1]
```diff
@@ -138,6 +139,40 @@ public void Deploy()
...
+                        totalWait.Stop();
+                        break;
+                    }
+
+                    WriteLog($"{remainingSilos} silos do not have a consistent cluster view, waiting until stabilization.");
+                    await Task.Delay(TimeSpan.FromMilliseconds(100));
+
+                    if (totalWait.Elapsed < TimeSpan.FromSeconds(60))
+                    {
+                        WriteLog($"Warning! {remainingSilos} silos do not have a consistent cluster view after {totalWait.ElapsedMilliseconds}ms, continuing without stabilization.");
```
Comment (xiazen) [inline src/Orleans.TestingHost/TestCluster.cs]: I think you may forget a break here. If while loop runs longer than a certain time, then break from the loop?

File: src/Orleans.TestingHost/TestCluster.cs [hunk 2]
```diff
@@ -138,6 +139,40 @@ public void Deploy()
                 string startMsg = "----------------------------- STARTING NEW UNIT TEST SILO HOST: " + GetType().FullName + " -------------------------------------";
                 WriteLog(startMsg);
                 await InitializeAsync();
+
+                // Poll each silo to check that it knows the expected number of active silos.
+                // If any silo does not have the expected number of active silos in its cluster membership oracle, try again.
+                // If the cluster membership has not stabilized after a certain period of time, give up and continue anyway.
+                var totalWait = Stopwatch.StartNew();
```
Comment (xiazen) [inline src/Orleans.TestingHost/TestCluster.cs]: can you capsulate this section of code into a function `WaitForClusterStabilizing` or `MakeSureAllSiloJoinedCluster` method. Just to make it cleaner and more readable

**Comment by sergeybykov** [comment]
Two tests are failing consistently:

**UnitTests.RuntimeCodeGenTests.RuntimeCodeGen_BasicEndToEnd**
>MESSAGE:
System.AggregateException : One or more errors occurred.\r\n---- System.InvalidOperationException : Cannot find generated GrainReference class for interface 'Orleans.Runtime.TestHooks.ITestHooksSystemTarget'\r\n---- The following constructor parameters did not have matching fixture data: Fixture fixture
+++++++++++++++++++
STACK TRACE:
----- Inner Stack Trace #1 (System.InvalidOperationException) ----- at Orleans.Runtime.TypeMetadataCache.GetGrainReferenceType(Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\AssemblyLoader\TypeMetadataCache.cs:line 53 at Orleans.GrainFactory.MakeCaster(Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 252 at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory) at Orleans.GrainFactory.Cast(IAddressable grain, Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 239 at Orleans.GrainFactory.Cast[TGrainInterface](IAddressable grain) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 223 at Orleans.GrainFactory.GetSystemTarget[TGrainInterface](GrainId grainId, SiloAddress destination) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 286 at Orleans.TestingHost.TestCluster.<DeployAsync>d__36.MoveNext() in D:\j\workspace\functional_pr---83d31c72\src\Orleans.TestingHost\TestCluster.cs:line 205 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Orleans.TestingHost.TestCluster.Deploy() in D:\j\workspace\functional_pr---83d31c72\src\Orleans.TestingHost\TestCluster.cs:line 126 at TestExtensions.BaseTestClusterFixture..ctor() in D:\j\workspace\functional_pr---83d31c72\test\TestExtensions\BaseClusterFixture.cs:line 43 ----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----

**UnitTests.RuntimeCodeGenTests.RuntimeCodeGen_AddsSupportClasses**
>System.AggregateException : One or more errors occurred.\r\n---- System.InvalidOperationException : Cannot find generated GrainReference class for interface 'Orleans.Runtime.TestHooks.ITestHooksSystemTarget'\r\n---- The following constructor parameters did not have matching fixture data: Fixture fixture
+++++++++++++++++++
STACK TRACE:
----- Inner Stack Trace #1 (System.InvalidOperationException) ----- at Orleans.Runtime.TypeMetadataCache.GetGrainReferenceType(Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\AssemblyLoader\TypeMetadataCache.cs:line 53 at Orleans.GrainFactory.MakeCaster(Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 252 at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory) at Orleans.GrainFactory.Cast(IAddressable grain, Type interfaceType) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 239 at Orleans.GrainFactory.Cast[TGrainInterface](IAddressable grain) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 223 at Orleans.GrainFactory.GetSystemTarget[TGrainInterface](GrainId grainId, SiloAddress destination) in D:\j\workspace\functional_pr---83d31c72\src\Orleans.Core\Core\GrainFactory.cs:line 286 at Orleans.TestingHost.TestCluster.<DeployAsync>d__36.MoveNext() in D:\j\workspace\functional_pr---83d31c72\src\Orleans.TestingHost\TestCluster.cs:line 205 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Orleans.TestingHost.TestCluster.Deploy() in D:\j\workspace\functional_pr---83d31c72\src\Orleans.TestingHost\TestCluster.cs:line 126 at TestExtensions.BaseTestClusterFixture..ctor() in D:\j\workspace\functional_pr---83d31c72\test\TestExtensions\BaseClusterFixture.cs:line 43 ----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----

**Comment by ReubenBond** [comment]
@dotnet-bot test bvt

We see this deps.json failure sometimes

**Pull request merged**